pkg/alertmanager/distributor.go

```diff
@@ -0,0 +1,237 @@
+
+	logger log.Logger
+}
+
+// NewDistributor constructs a new Distributor
+func NewDistributor(cfg *MultitenantAlertmanagerConfig, alertmanagersRing *ring.Ring, alertmanagerClientsPool ClientsPool, logger log.Logger, reg prometheus.Registerer) (d *Distributor, err error) {
+	if alertmanagerClientsPool == nil {
+		alertmanagerClientsPool = newAlertmanagerClientsPool(client.NewRingServiceDiscovery(alertmanagersRing), cfg.AlertmanagerClient, logger, reg)
+	}
+
+	d = &Distributor{
+		cfg:                     cfg,
+		logger:                  logger,
+		alertmanagerRing:        alertmanagersRing,
+		alertmanagerClientsPool: alertmanagerClientsPool,
+	}
+
+	d.Service = services.NewBasicService(nil, d.running, nil)
+	return d, nil
+}
+
+func (d *Distributor) running(ctx context.Context) error {
+	<-ctx.Done()
+	d.requestsInFlight.Wait()
+	return nil
+}
+
+// IsPathSupported returns true if the given route is currently supported by the Distributor.
+func (d *Distributor) IsPathSupported(path string) bool {
+	// API can be found at https://petstore.swagger.io/?url=https://raw.githubusercontent.com/prometheus/alertmanager/master/api/v2/openapi.yaml.
+	return strings.HasSuffix(path, "/alerts") ||
+		strings.HasSuffix(path, "/alerts/groups")
+}
+
+// DistributeRequest shards the writes and returns as soon as the quorum is satisfied.
+// In case of reads, it proxies the request to one of the alertmanagers.
+// DistributeRequest assumes that the caller has verified IsPathSupported return
+// true for the route.
+func (d *Distributor) DistributeRequest(w http.ResponseWriter, r *http.Request) {
+	d.requestsInFlight.Add(1)
+	defer d.requestsInFlight.Done()
+
+	userID, err := tenant.TenantID(r.Context())
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusUnauthorized)
+		return
+	}
+
+	logger := log.With(d.logger, "user", userID)
+
+	if r.Method == http.MethodGet || r.Method == http.MethodHead {
+		d.doRead(userID, w, r, logger)
+	} else {
+		d.doWrite(userID, w, r, logger)
+	}
+}
+
+func (d *Distributor) doWrite(userID string, w http.ResponseWriter, r *http.Request, logger log.Logger) {
+	var body []byte
+	var err error
+	if r.Body != nil {
+		body, err = ioutil.ReadAll(r.Body)
+		if err != nil {
+			if err.Error() == "http: request body too large\n" {
+				http.Error(w, "Request body too large", http.StatusBadRequest)
+			}
+			level.Error(logger).Log("msg", "failed to read the request body during write", "err", err)
+			w.WriteHeader(http.StatusInternalServerError)
+			return
+		}
+	}
+
+	var firstSuccessfulResponse *httpgrpc.HTTPResponse
+	var respMtx sync.Mutex
+	grpcHeaders := httpTogrpchttpHeaders(r.Header)
+	err = ring.DoBatch(r.Context(), RingOp, d.alertmanagerRing, []uint32{shardByUser(userID)}, func(am ring.InstanceDesc, _ []int) error {
+		// Use a background context to make sure all alertmanagers get the request even if we return early.
+		localCtx, cancel := context.WithTimeout(context.Background(), d.cfg.AlertmanagerClient.RemoteTimeout)
```

Comment:
Why do you need it? I understand the comment, but I can't understand who will cancel the context if "we return early".

Reply:
This logic is copied from the pkg/distributor code. So you mean not required here?